framework/src/wifi_manager/wifi_manager.c

```diff
@@ -489,284 +590,353 @@ wifi_manager_result_e wifi_manager_init(wifi_manager_cb_s *wmcb)
-		ndbg("w_info_mutex_destroy fail %d", result);
-	}
 
-error_without_mutex_release:
-	if (w_mutex != NULL) {
-		free(w_mutex);
-		w_mutex = NULL;
-	}
-	if (w_info_mutex != NULL) {
-		free(w_info_mutex);
-		w_info_mutex = NULL;
+wifi_manager_result_e _handle_disconnecting_state(_wifimng_msg_s *msg)
+{
+	WM_LOG_HANDLER_START;
+	if (msg->event == EVT_STA_DISCONNECTED) {
+		if (WIFIMNG_CHECK_CBK) {
+			_handle_user_cb(CB_EVT_STA_DISCONNECT, NULL);
+		} else {
+			WIFIMNG_RESET_CBK_CHK;
+		}
+		WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTED);
+	} else {
+		return WIFI_MANAGER_FAIL;
 	}
-
-	return WIFI_MANAGER_FAIL;
+	return WIFI_MANAGER_SUCCESS;
 }
 
-wifi_manager_result_e wifi_manager_deinit()
+
+wifi_manager_result_e _handle_connecting_state(_wifimng_msg_s *msg)
 {
-	CHECK_WITH_RETURN("deinit");
+	WM_LOG_HANDLER_START;
+	wifi_manager_result_e res;
+	if (msg->event == EVT_STA_CONNECTED) {
+		res = _get_ipaddr_dhcp();
+		if (res != WIFI_MANAGER_SUCCESS) {
+			_handle_user_cb(CB_EVT_STA_CONNECT_FAIL, NULL);
+			res = _wifimng_disconnect_ap(); // without callback
+			if (res != WIFI_MANAGER_SUCCESS) {
+				// this should succeed
+			}
+			WIFIMNG_SET_NO_CBK;
+			WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTING);
+			return WIFI_MANAGER_FAIL;
+		}
+		_handle_user_cb(CB_EVT_STA_CONNECTED, NULL);
+		WIFIMNG_SET_STATE(WIFIMNG_STA_CONNECTED);
+	} else if (msg->event == EVT_STA_CONNECT_FAIL) {
+		_handle_user_cb(CB_EVT_STA_CONNECT_FAIL, NULL);
+		WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTED);
+	} else {
+		return WIFI_MANAGER_FAIL;
+	}
+	return WIFI_MANAGER_SUCCESS;
+}
 
-	wifi_utils_result_e result = WIFI_UTILS_SUCCESS;
-	wifi_manager_result_e ret = WIFI_MANAGER_SUCCESS;
 
-	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
-	/* wifi_manager mode is switched to wifi_deinitializing */
-	if ((g_manager_info.mode != WIFI_INITIALIZING) && (g_manager_info.mode != WIFI_MODE_CHANGING) && \
-			(g_manager_info.mode != WIFI_DEINITIALIZING)) {
-		g_manager_info.pre_mode = g_manager_info.mode;
-		g_manager_info.next_mode = WIFI_NONE;
-		g_manager_info.mode = WIFI_DEINITIALIZING;
+wifi_manager_result_e _handle_connected_state(_wifimng_msg_s *msg)
+{
+	WM_LOG_HANDLER_START;
+	wifi_manager_result_e res;
+	if (msg->event == EVT_DISCONNECT) {
+		// call disconnect_ap
+		res = _wifimng_disconnect_ap();
+		if (res != WIFI_MANAGER_SUCCESS) {
+			// this should succeed
+		}
+		WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTING);
+	} else if (msg->event == EVT_STA_DISCONNECTED) {
+		// ap is disconnected from remote
+		_handle_user_cb(CB_EVT_STA_DISCONNECT, NULL);
+		WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTED);
+	} else if (msg->event == EVT_SET_SOFTAP) {
+		res = _wifimng_connect_ap((wifi_manager_ap_config_s *)msg->param);
+		if (res != WIFI_MANAGER_SUCCESS) {
+			res = _wifimng_disconnect_ap();
+			if (res != WIFI_MANAGER_SUCCESS) {
+				// this should succeed
+			}
+			WIFIMNG_SET_STATE(WIFIMNG_STA_DISCONNECTING);
+		}
+		WIFIMNG_SET_STATE(WIFIMNG_SOFTAP);
```

Comment:
There is no function call to set softAP.

Reply:
it is but I will fix